### PR TITLE
fix reading data sources that include MS SQL Server queries

### PR DESCRIPTION
**Pull request by doemming**

### CODE REVIEW

File: io/plugins/eu.esdihumboldt.hale.io.jdbc/src/eu/esdihumboldt/hale/io/jdbc/JDBCTableCollection.java [hunk 1]
```diff
@@ -242,8 +242,13 @@ public JDBCTableCollection(TypeDefinition type, URI jdbcURI, String user, String
...
 			// support project variables
 			query = JDBCUtil.replaceVariables(query, services);
-
-			this.countQuery = null;
+			
+			// this.countQuery = null; 
+			// countQuery = null caused to return UNKNOWN_SIZE in size() with causes isEmpty() 
+			// to return false and that causes problems on iterating with MultiInstanceCollection 
+			this.countQuery = "SELECT COUNT(*) FROM (\n" + query + "\n) tmp";
+			// note 1: this sub query is not supported in all SQL dialects
```
Comment (stempler) [inline io/plugins/eu.esdihumboldt.hale.io.jdbc/src/eu/esdihumboldt/hale/io/jdbc/JDBCTableCollection.java:250]: Do you have an idea in which it is not supported?

File: common/plugins/eu.esdihumboldt.hale.common.instance/src/eu/esdihumboldt/hale/common/instance/model/impl/MultiInstanceCollection.java [hunk 2]
```diff
@@ -194,6 +194,7 @@ public Instance next() {
 						currentIterator.close();
 					currentCollection = hasNextIndex;
 					currentIterator = collections.get(currentCollection).iterator();
+					currentIterator.hasNext(); // force to initialize currentIterator.hasNext flag
```
Comment (stempler) [inline common/plugins/eu.esdihumboldt.hale.common.instance/src/eu/esdihumboldt/hale/common/instance/model/impl/MultiInstanceCollection.java]: What difference does it make?
My feeling is that if this is necessary we should try to fix the cause and not the symptom.

File: io/plugins/eu.esdihumboldt.hale.io.jdbc/src/eu/esdihumboldt/hale/io/jdbc/JDBCTableCollection.java (hunk 1 above)
Comment (doemming) [reply io/plugins/eu.esdihumboldt.hale.io.jdbc/src/eu/esdihumboldt/hale/io/jdbc/JDBCTableCollection.java:250]: I am not sure if subquery is part of standard SQL syntax. However accoring to a quick research all major databases seem to support.

File: common/plugins/eu.esdihumboldt.hale.common.instance/src/eu/esdihumboldt/hale/common/instance/model/impl/MultiInstanceCollection.java (hunk 2 above)
Comment (doemming) [reply common/plugins/eu.esdihumboldt.hale.common.instance/src/eu/esdihumboldt/hale/common/instance/model/impl/MultiInstanceCollection.java]: You are right. With the change in https://github.com/stempler/hale/pull/3/files/499d2980a4f7f6bbaaae954abacf5027e0a1966a#diff-5ddc8e3e73548a88d2325686f278149f
this change is not required anymore.

Comment (stempler) [reply common/plugins/eu.esdihumboldt.hale.common.instance/src/eu/esdihumboldt/hale/common/instance/model/impl/MultiInstanceCollection.java]: Ok, so you remove this and then we merge it?

**Pull request merged**